kds/kds.go

```diff
@@ -227,33 +323,104 @@ func ComposeTCBParts(parts TCBParts) (TCBVersion, error) {
 		check127("Spl4", parts.Spl4),
 		check127("TeeSpl", parts.TeeSpl),
 		check127("BlSpl", parts.BlSpl),
+		check127("FmcSpl", parts.FmcSpl),
 	); err != nil {
-		return TCBVersion(0), err
+		return tcbVersion, err
 	}
-	return TCBVersion(
-		(uint64(parts.UcodeSpl) << 56) |
+
+	switch parts.version {
+	case TCBStructVersion0:
+		tcbVersion.version = TCBStructVersion0
+		tcbVersion.TCB = (uint64(parts.UcodeSpl) << 56) |
 			(uint64(parts.SnpSpl) << 48) |
 			(uint64(parts.Spl7) << 40) |
 			(uint64(parts.Spl6) << 32) |
 			(uint64(parts.Spl5) << 24) |
 			(uint64(parts.Spl4) << 16) |
 			(uint64(parts.TeeSpl) << 8) |
-			(uint64(parts.BlSpl) << 0)), nil
+			(uint64(parts.BlSpl) << 0)
+	case TCBStructVersion1:
+		tcbVersion.version = TCBStructVersion1
+		tcbVersion.TCB = (uint64(parts.UcodeSpl) << 56) |
+			(uint64(parts.SnpSpl) << 24) |
+			(uint64(parts.TeeSpl) << 16) |
+			(uint64(parts.BlSpl) << 8) |
+			(uint64(parts.FmcSpl) << 0)
+	default:
+		return tcbVersion, fmt.Errorf("unsupported TCB struct version: %v", parts.version)
+	}
+
+	return tcbVersion, nil
+}
+
+func (parts TCBParts) String() string {
+	switch parts.version {
+	case TCBStructVersion1:
+		return fmt.Sprintf("{BlSpl:%d TeeSpl:%d Spl5:%d Spl6:%d Spl7:%d SnpSpl:%d UcodeSpl:%d FmcSpl: %d}",
+			parts.BlSpl, parts.TeeSpl, parts.Spl5, parts.Spl6, parts.Spl7, parts.SnpSpl, parts.UcodeSpl, parts.FmcSpl)
+	default:
+		return fmt.Sprintf("{BlSpl:%d TeeSpl:%d Spl4:%d Spl5:%d Spl6:%d Spl7:%d SnpSpl:%d UcodeSpl:%d}",
+			parts.BlSpl, parts.TeeSpl, parts.Spl4, parts.Spl5, parts.Spl6, parts.Spl7, parts.SnpSpl, parts.UcodeSpl)
+	}
+}
+
+// ComposeTCBParts returns an SEV-SNP TCB_VERSION from OID mapping values. The spl4-spl7 fields are
+// reserved, but the KDS specification designates them as 4 byte-sized fields.
+//
+// Deprecated: use ComposeTCBPartsToTCBVersionStruct instead
+func ComposeTCBParts(parts TCBParts) (TCBVersion, error) {
+	tcb, err := parts.ToTCBVersionStruct()
+	if err != nil {
+		return TCBVersion(0), err
+	}
+
+	return TCBVersion(tcb.TCB), nil
+}
+
+// ComposeTCBPartsToTCBVersionStruct returns an SEV-SNP TCB_VERSION from OID mapping
+// values as a TCBVersionStruct. The spl4-spl7 fields are reserved, but the
+// KDS specification designates them as 4 byte-sized fields.
+func ComposeTCBPartsToTCBVersionStruct(parts TCBParts) (TCBVersionStruct, error) {
+	return parts.ToTCBVersionStruct()
```

Comment:
Why is this function necessary if you're exporting ToTCBVersionStruct?

Reply:
Sorry about that, I meant to remove it.